### ISSUE TITLE
Stream of payloads

### ISSUE THREAD

**Issue by Chris00** [inhabitedtype/angstrom#40]
I am parsing a protocol that is a stream of "frames", some of them containing a payload (the length of which is given in the frame "header") that I must pass down to users of the protocol.  It is easy enough to extract the payload as a string and pass it down. Each of the payload is not big (max 64KB), but the total size of the user data may in some cases be several GB.  Is there a more efficient way to pass this data around — say as some "stream" of `Cstruct.t` — that is/will be supported by Ångström?

**Comment by seliopou** [comment]
Is your question referring to support for a stream-like structure on the input side or the output side? If it's the former then the answer is: not currently, but I would like angstrom to support something like that in the future. If it's the latter, then the answer is: yes. Here's a bit more detail. Let me know if this answers your question.

## Output Streams

First, the latter case of generating an output stream. Supposing you have a frame parser and a function to push a frame onto some queue for later processing, you can construct a parser that will populate your frame queue like this:

```ocaml
(* val frame : frame Angstrom.t *)
(* val push : frame -> unit *)
let p : unit Angstrom.t = skip_many (frame >>| push <* commit)
```
The parser `p` will parse a `frame`, push it onto your queue, and then `commit` to the current parse, preventing backtracking to a position prior to the current position. Now, when you run the parser and receive a `Partial` state, you have an opportunity to process the contents of your frame queue before continuing the parse. You don't have to wait for the parser to run over the entire input in order to surface frames to the application for processing.

This is essentially how the `parse_many` functions are implemented in the `Angstrom_lwt_unix` and `Angstrom_async` modules. If you're using these modules, you won't have to worry about explicit scheduling.

## Input Streams

To support the former case of a stream of input chunks, angstrom needs a bit of work. It also needs to maintain a clear separation between parsing logic and IO operations. A lot of other parser combinator libraries use stream datatypes to do something along the lines of incremental input. That unfortunately forces blocking `read()` calls to occur in the innards of the parser, which is something that angstrom must avoid at all costs.

But _something_ like an input stream must be supported by the library, if for no other reason than to support vectorized IO, which will further minimize copies. I think the best solution would be to change the input type from this:

```ocaml
type input =
  [ `String    of string
  | `Bigstring of bigstring ]
```

to something like this:

```ocaml
type 'a iovec =
  { buffer : 'a
  ; off : int
  ; len : int }

type input = [ `String of string | `Bigstring of bigstring ] iovec list
```

In this world, client code that's managing an input stream could collect what is immediately available from the stream and pass it all into the parser. When the `Parser` reaches a partial state, it can shift the previous input based on the `partial.consumed` field, append any newly-available chunks to the input, and pass it back to the parser for further processing.

This approach has the benefit of supporting both vectorized IO and a stream-like input model while minimizing copies in both situations. It also doesn't run afoul of cooperative threading to achieve it. The one downside is that the implementation may become a bit more complex and possibly slower, now that's necessary to index into a list (and to determine that index) to retrieve a byte of input.

**Comment by Chris00** [comment]
Thanks for your answer.  It is mostly the first part that I am interested into right now but the second part is interesting as well.

I still miss a piece of information though.  The frames have a header so the reading part will look like:
```ocaml
header >>= fun h → frame h >>| push h <* commit
```
The frame is really nothing else than a sequence of bytes of some length given in `h`.  If, say, the underlying input is a bigstring, is there a way to `push` the slice of bytes without copying (say with `Bigstring.sub_shared`).  Do I have to use the *private* `pos` function or is it possible without that?

**Comment by seliopou** [comment]
I get your meaning now. At the moment there isn't a way to surface bytes from the input without a copy, but I definitely want to include this functionality, somehow. I contemplated an API that would allow for this before angstrom's initial release, but I decided against including it simply because I didn't have an application for the API to test and design against. Maybe it's as simple as including a parser like this one:

```ocaml
val unsafe_take_bigstring : int -> bigstring t
(** [unsafe_take_bigstring n] accepts exactly [n] characters of input and
    returns them as a bigstring. When possible, the bigstring will be a view into
    the current input and should therefore not be modified.  *)
```

If you have any other suggestions, I'm all ears.

**Comment by Chris00** [comment]
You may want to use a `Cstruct.t` to return the slice of your bigarray (`Array1.sub` is more expensive).  Also, returning a list of chunks—instead of a single large one—will accommodate for inputs that come in several pieces (such as those coming from ``read : flow → [`Ok of Cstruct.t | ...]`` in Mirage):
```ocaml
val unsafe_take : int → Cstruct.t list t
```

**Comment by seliopou** [comment]
Returning the list will have to wait until the input stream actually accepts a list of chunks. I'm tempted to switch the interface so that the `Unbuffered` module only takes `Cstruct.t`'s as input, and the `Buffered` module can take a `string`, `bigstring`, or `Cstruct.t`, all of which are copied. Given that, I think that `unsafe_take` will be a bit more sensible.

**Comment by seliopou** [comment]
This was actually implemented in #99.
